test/widgets/app_test.dart

```diff
@@ -57,6 +59,106 @@ void main() {
     });
   });
 
+  group('_PreventEmptyStack', () {
+    late List<Route<void>> pushedRoutes;
+    late List<Route<void>> removedRoutes;
+    late List<Route<void>> poppedRoutes;
+
+    Future<void> prepare(WidgetTester tester) async {
+      addTearDown(testBinding.reset);
+
+      pushedRoutes = [];
+      removedRoutes = [];
+      poppedRoutes = [];
+      final testNavObserver = TestNavigatorObserver();
+      testNavObserver.onPushed = (route, prevRoute) => pushedRoutes.add(route);
+      testNavObserver.onRemoved = (route, prevRoute) => removedRoutes.add(route);
+      testNavObserver.onPopped = (route, prevRoute) => poppedRoutes.add(route);
+
+      await tester.pumpWidget(ZulipApp(navigatorObservers: [testNavObserver]));
+      await tester.pump(); // start to load account
+      check(pushedRoutes).single.isA<WidgetRoute>().page.isA<HomePage>();
+      pushedRoutes.clear();
+    }
+
+    testWidgets('do not push route to non-empty navigator stack', (tester) async {
```

Comment:
nit: I think the narrative would be clearer if these three test cases:

```dart
    testWidgets('do not push route to non-empty navigator stack', (tester) async {
    // …
    testWidgets('push route when popping last route on stack', (tester) async {
    // …
    testWidgets('push route when removing last route on stack', (tester) async {
```

appeared in the opposite order.

In particular the "non-empty stack" case is a variation on the "popping last route" case, with some extra wrinkles added. So it's clearest to understand when taking that other case as a point of departure.